pkg/cloudprovider/aws/fleet/packing/packing.go

```diff
@@ -38,19 +42,141 @@ type Packings struct {
+	// try to fit as many pods as possible. return the list of instances with
+	// highest pods packed and the pods.
+
+	// TODO add filters
+	// TODO reserve (Kubelet+ daemon sets) overhead for instance types
+	// TODO count number of pods created on an instance type
+	for _, instanceOption := range p.getInstanceTypes("") {
+		// check how many pods we can fit on this instance type
+		packings, err := p.calculatePackingForInstance(instanceOption, pods)
+		if err != nil {
+			zap.S().Errorf("Failed to calculate packing for instance type %s, err, %w", instanceOption.instanceType, err)
+			continue
+		}
+		if len(packings) == 0 {
+			continue
+		}
+		// If the pods packed are the same as before, this instance type can be
+		// considered as a backup option in case we get ICE
+		if podsMatch(previousPacking, packings) {
+			instanceTypesSelected = append(instanceTypesSelected, instanceOption)
+		} else if len(packings) > len(previousPacking) {
+			// If pods packed are more than compared to what we got in last
+			// iteration, consider using this instance type
+			previousPacking = packings
+			instanceTypesSelected = []*nodeCapacity{instanceOption}
+		}
+	}
+	instanceTypeNames := []string{}
+	for _, instance := range instanceTypesSelected {
+		instanceTypeNames = append(instanceTypeNames, instance.instanceType)
+	}
+	return previousPacking, instanceTypeNames
+}
+
+func (p *packingEstimator) calculatePackingForInstance(instance *nodeCapacity, podList []*v1.Pod) ([]*v1.Pod, error) {
+	packing := []*v1.Pod{}
+	// start with the smallest pod based on resources requested
+	for _, pod := range podList {
+		if p.isPodPacked[pod] {
+			continue
+		}
+		cpu := calculateCPURequested(pod)
+		memory := calculateMemoryRequested(pod)
+		if err := instance.reserveCapacity(cpu, memory); err != nil {
+			if errors.Is(err, ErrInsufficientCapacity) {
+				// First we need to find the instance on which we can fit the
+				// largest pods, it can't fit here, return and try with a next
+				// bigger instance type
+				if len(packing) == 0 {
+					return packing, nil
+				}
+				// We have already packed a bigger pod on this instance type, if
+				// we can't pack this pod, check the next pod it might be
+				// smaller in size and help fill the gaps on the instance
+				continue
+			}
+			return nil, fmt.Errorf("reserve capacity failed %w", err)
+		}
+		packing = append(packing, pod)
+	}
+	instance.utilizedCapacity = v1.ResourceList{
+		v1.ResourceCPU:    resource.MustParse("0"),
+		v1.ResourceMemory: resource.MustParse("0"),
+	}
+	return packing, nil
+}
+
+func podsMatch(first, second []*v1.Pod) bool {
+	if len(first) != len(second) {
+		return false
+	}
+	podkey := func(pod *v1.Pod) string {
+		return pod.Namespace + "/" + pod.Name
+	}
+	podSeen := map[string]struct{}{}
+	for _, pod := range first {
+		podSeen[podkey(pod)] = struct{}{}
+	}
+	for _, pod := range second {
+		if _, ok := podSeen[podkey(pod)]; !ok {
+			return false
+		}
+	}
```

Comment:
I don't think this handles the case like this:

```
first = {p1,p2}
second = {p1,p1}
```

? I think that's why this test code in the kubelet does it slightly differently:

```go
// podListEqual checks if the lists contents contain all of the same elements.
func podListEqual(list1 []*v1.Pod, list2 []*v1.Pod) bool {
	if len(list1) != len(list2) {
		return false
	}

	m := map[*v1.Pod]int{}
	for _, val := range list1 {
		m[val] = m[val] + 1
	}
	for _, val := range list2 {
		m[val] = m[val] - 1
	}
	for _, v := range m {
		if v != 0 {
			return false
		}
	}
	return true
}
```